### ISSUE TITLE
Misleading error message, privacy error reported as type error

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#64684]
```rust
mod foo {
    pub struct Foo {
        private: i32,
    }
}

use foo::Foo;

fn main() {
    let mut foo = Foo { private: 0_usize };
}

```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=db546b8ab3361a008f7cd421658e15aa))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
  --> src/main.rs:11:34
   |
11 |     let mut foo = Foo { private: 0_usize } ;
   |                                  ^^^^^^^ expected i32, found usize

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

The error message here is incorrect. Whether the types match or not is irrelevant, because the field is private. The error should say something like this instead ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c2a74939542ed8ce8535b779a2688c1c)): 

```
 error[E0451]: field `private` of struct `foo::Foo` is private
  --> src/main.rs:10:25
   |
10 |     let mut foo = Foo { private: 0_i32 };
   |                         ^^^^^^^^^^^^^^ field `private` is private

error: aborting due to previous error
```

**Comment by TheSamsa** [comment]
I would like to tackle this as a first timer, if that is okay?
A little pointer where to begin my trip would be awesome, thanks in advance!

**Comment by gnzlbg** [comment]
cc @estebank

**Comment by witling** [comment]
> 
> 
> I would like to tackle this as a first timer, if that is okay and not too overly complicated for a first time issue ;)
> A little pointer where to begin my trip would be awesome, thanks in advance!

I don't believe that this is a good first time issue. Type checking is done before executing privacy checks. ([see here](https://github.com/rust-lang/rust/blob/ada102456de38e462b4f00473dd923f15dc3f296/compiler/rustc_interface/src/passes.rs#L816))
Changing the order seems to have a negative performance impact in a lot of cases as type errors seem to be way more common. At least that is what my 20 minute code skim came up to.

**Comment by TheSamsa** [comment]
> > I would like to tackle this as a first timer, if that is okay and not too overly complicated for a first time issue ;)
> > A little pointer where to begin my trip would be awesome, thanks in advance!
> 
> I don't believe that this is a good first time issue. Type checking is done before executing privacy checks. ([see here](https://github.com/rust-lang/rust/blob/ada102456de38e462b4f00473dd923f15dc3f296/compiler/rustc_interface/src/passes.rs#L816))
> Changing the order seems to have a negative performance impact in a lot of cases as type errors seem to be way more common. At least that is what my 20 minute code skim came up to.

Hi thank you for your insights!
I won't do it then, since my time is in short supply right now. It is better for someone else to step in.
